### ISSUE TITLE
Fail to build with a failed condition in a UDC

### ISSUE THREAD

**Issue by mvgijssel** [earthly/earthly#1417]
I have a curious problem where if an Earthfile combines a UDC, a failed condition a `SAVE IMAGE` with a `--cache-from` statement inside the `WITH DOCKER` statement it results in the following error:

```
loading config values from "./earthly_config.yml"
           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)
ubuntu:focal@sha256:626ffe58f6e7566e00254b638eb7e0f3b11d4da9675088f4781a50ae288f3322 | --> Load metadata linux/amd64
 earthly/dind:alpine | --> Load metadata linux/amd64
               +base | --> FROM ubuntu:focal@sha256:626ffe58f6e7566e00254b638eb7e0f3b11d4da9675088f4781a50ae288f3322
               +base | [██████████] resolve docker.io/library/ubuntu:focal@sha256:626ffe58f6e7566e00254b638eb7e0f3b11d4da9675088f4781a50ae288f3322 ... 100%
              +image | *cached* --> IF [ 1 = 2 ]
               cache | --> importing cache manifest from my_test_image:master
Error stack trace:
github.com/moby/buildkit/util/stack.Enable
	github.com/moby/buildkit@v0.8.2-0.20210129065303-6b9ea0c202cf/util/stack/stack.go:77
github.com/moby/buildkit/util/grpcerrors.FromGRPC
	github.com/moby/buildkit@v0.8.2-0.20210129065303-6b9ea0c202cf/util/grpcerrors/grpcerrors.go:188
github.com/moby/buildkit/util/grpcerrors.UnaryClientInterceptor
	github.com/moby/buildkit@v0.8.2-0.20210129065303-6b9ea0c202cf/util/grpcerrors/intercept.go:41
google.golang.org/grpc.(*ClientConn).Invoke
	google.golang.org/grpc@v1.38.0/call.go:35
github.com/moby/buildkit/api/services/control.(*controlClient).Solve
	github.com/moby/buildkit@v0.8.2-0.20210129065303-6b9ea0c202cf/api/services/control/control.pb.go:1321
github.com/moby/buildkit/client.(*Client).solve.func2
	github.com/moby/buildkit@v0.8.2-0.20210129065303-6b9ea0c202cf/client/solve.go:215
golang.org/x/sync/errgroup.(*Group).Go.func1
	golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57
runtime.goexit
	runtime/asm_amd64.s:1371
Error: build target: build main: bkClient.Build: failed to solve: Canceled: context canceled
```

The build passes in any of the following circumstances:
- If I build the `+image` directly
- if I move the `SAVE IMAGE` outside of the UDC
- if I remove the condition
- if I make the condition pass (replace 1 = 2 with 1 = 1)
- if I remove the `--cache-from` argument

The Earthfile is as follows:

```dockerfile
FROM ubuntu:focal@sha256:626ffe58f6e7566e00254b638eb7e0f3b11d4da9675088f4781a50ae288f3322 

image:
    DO ./tools/earthly+SAVE_IMAGE

test:
    FROM earthly/dind:alpine
    WITH DOCKER \
        --load my_image:dev=./+image

        RUN --mount type=secret,target=/root/meltano_secrets,id=+secrets/meltano_secrets \
            docker run my_image:dev /bin/bash -c "echo Hello World!"
    END
```

the UDC:

```dockerfile
SAVE_IMAGE:
    COMMAND

    IF [ 1 = 2 ]
        RUN echo "false"
    END

    SAVE IMAGE --cache-from my_test_image:master --push my_test_image:dev
```

My environment:

```
earthly version v0.5.24 darwin/amd64; macOS
Docker version 20.10.8, build 3967b7d
```

Earthly config

```
global:
  disable_analytics: true
  cache_size_mb: 70000
  buildkit_additional_args:
    [
      "-e",
      "BUILDKIT_STEP_LOG_MAX_SIZE=-1",
      "-e",
      "BUILDKIT_STEP_LOG_MAX_SPEED=-1",
    ]
```

**Comment by andrewsykim** [comment]
Thanks for detailed bug report @mvgijssel!

I'm going to attempt to reproduce this issue. Were there any environment variables being set in addition to the earthly config you shared?

**Comment by mvgijssel** [comment]
Yeah sure!

```bash
EARTHLY_BUILDKIT_IMAGE=earthly/buildkitd:v0.5.24
EARTHLY_BUILD_ARGS=CI=false,REGISTRY_DOCKERHUB=docker.io
EARTHLY_CI=false
EARTHLY_CONFIG=./earthly_config.yml
EARTHLY_INTERACTIVE=true
EARTHLY_NO_OUTPUT=true
EARTHLY_SECRET_FILES=secrets=.env
```

**Comment by vladaionescu** [comment]
I believe that @alexcb has fixed this and it has already been released in Earthly `v0.6`

**Comment by alexcb** [comment]
This was (temporarily) fixed with https://github.com/earthly/buildkit/commit/5ee328821a2b2b3cbb84ef6290ba8d1abc910e44

Ideally we'll figure out the root cause and move the call to `.wait()` somewhere more appropriate. In either case, this issue should no longer occur.
